packages/protocol-kit/src/utils/transactions/gas.ts

```diff
@@ -151,3 +189,285 @@ export async function estimateTxGas(
+ *
+ * The estimation method differs based on the version of the Safe:
+ * - For versions >= 1.3.0, the simulate function defined in the simulateTxAccessor Contract is used.
+ * - For versions < 1.3.0, the deprecated requiredTxGas method defined in the GnosisSafe contract is used.
+ *
+ * @async
+ * @function estimateSafeTxGas
+ * @param {Safe} safe - The Safe instance containing all the necessary information about the safe.
+ * @param {SafeTransaction} safeTransaction - The transaction for which the safeTxGas is to be estimated.
+ * @returns {Promise<string>} A Promise that resolves with the estimated safeTxGas as a string.
+ *
+ */
+export async function estimateSafeTxGas(
+  safe: Safe,
+  safeTransaction: SafeTransaction
+): Promise<string> {
+  const safeVersion = await safe.getContractVersion()
+
+  if (semverSatisfies(safeVersion, '>=1.3.0')) {
+    return estimateSafeTxGasWithSimulate(safe, safeTransaction)
+  } else {
+    return estimateSafeTxGasWithRequiredTxGas(safe, safeTransaction)
+  }
+}
+
+/**
+ * This function estimates the safeTxGas of a Safe transaction.
+ * Using the deprecated method of requiredTxGas defined in the GnosisSafe contract. This method is meant to be used for Safe versions < 1.3.0.
+ *
+ * @async
+ * @function estimateSafeTxGasWithRequiredTxGas
+ * @param {Safe} safe - The Safe instance containing all the necessary information about the safe.
+ * @param {SafeTransaction} safeTransaction - The transaction for which the safeTxGas is to be estimated.
+ * @returns {Promise<string>} A Promise that resolves with the estimated safeTxGas as a string.
+ *
+ */
+async function estimateSafeTxGasWithRequiredTxGas(
+  safe: Safe,
+  safeTransaction: SafeTransaction
+): Promise<string> {
+  const isSafeDeployed = await safe.isSafeDeployed()
+  const safeAddress = await safe.getAddress()
+  const safeVersion = await safe.getContractVersion()
+  const ethAdapter = safe.getEthAdapter()
+  const safeSingletonContract = await getSafeContract({
+    ethAdapter,
+    safeVersion
+  })
+
+  // deprecated method to estimate the safeTxGas of a Safe transaction, see: https://github.com/safe-global/safe-contracts/blob/v1.2.0/contracts/GnosisSafe.sol#L276
+  const transactionDataToEstimate: string = safeSingletonContract.encode('requiredTxGas', [
+    safeTransaction.data.to,
+    safeTransaction.data.value,
+    safeTransaction.data.data,
+    safeTransaction.data.operation
+  ])
+
+  const to = isSafeDeployed ? safeAddress : safeSingletonContract.getAddress()
+
+  const transactionToEstimateGas = {
+    to,
+    value: '0',
+    data: transactionDataToEstimate,
+    from: safeAddress
+  }
+  try {
+    const encodedResponse = await ethAdapter.call(transactionToEstimateGas)
+
+    const safeTxGas = '0x' + encodedResponse.slice(-32)
+
+    return safeTxGas
+
+    // if the call throws an error we try to parse the returned value
+  } catch (error: any) {
```

Comment:
Cannot be typed?

Reply:
The error depends on which Ethereum provider you are using. That's why we need to use `any` and try to parse the returned value differently for each case, whether it's `Web3` or `Ethers`.